### ISSUE TITLE
Bug in pocl_basic_copy

### ISSUE THREAD

**Issue by opti-mix** [pocl/pocl#677]
The current version of `pocl_basic_copy` is defined as:
```c
void
pocl_basic_copy (void *data,
                 pocl_mem_identifier * dst_mem_id,
                 cl_mem dst_buf,
                 pocl_mem_identifier * src_mem_id,
                 cl_mem src_buf,
                 size_t dst_offset,
                 size_t src_offset,
                 size_t size)
{
  void *__restrict__ src_ptr = src_mem_id->mem_ptr;
  void *__restrict__ dst_ptr = dst_mem_id->mem_ptr;
  if (src_ptr  == dst_ptr)
    return;

  memcpy ((char *)dst_ptr + dst_offset, (char *)src_ptr + src_offset, size);
}
```
But if both copy is performed for the different regions of the same buffer, the `if` condition would be `true` and no copying would be performed at all! This is clearly a bug.

The fix seems pretty obvious. One needs to compare the pointers taking the offsets into account, e.g. as in the following code:

```c
void
pocl_basic_copy (void *data,
                 pocl_mem_identifier * dst_mem_id,
                 cl_mem dst_buf,
                 pocl_mem_identifier * src_mem_id,
                 cl_mem src_buf,
                 size_t dst_offset,
                 size_t src_offset,
                 size_t size)
{
  void *__restrict__ src_ptr = src_mem_id->mem_ptr;
  void *__restrict__ dst_ptr = dst_mem_id->mem_ptr;
  if ((char *)src_ptr + src_offset == (char *)dst_ptr + dst_offset)
    return;

  memcpy ((char *)dst_ptr + dst_offset, (char *)src_ptr + src_offset, size);
}
```

**Comment by franz** [comment]
Hello. This bug was actually fixed recently in git master, by commit 5a45af695aeb4a. Please retry with git master

**Comment by opti-mix** [comment]
> This bug was actually fixed recently in git master, by commit 5a45af6. Please retry with git master

Confirmed! Tested with git master and it is fixed. Thanks!
